File: models/mlb/cartridges/MLB-RP36/check_run.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

from snapshot_run import ROOT, RUN_ROOT, build_snapshot


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Check an MLB-RP36 reliever addendum run snapshot.")
    parser.add_argument("--date", required=True, help="Slate date, YYYY-MM-DD.")
    return parser.parse_args()


def read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def first_diff(left: Any, right: Any, prefix: str = "$") -> str | None:
    if left == right:
        return None
    if type(left) is not type(right):
        return f"{prefix}: type {type(left).__name__} != {type(right).__name__}"
    if not isinstance(left, (dict, list)):
        return f"{prefix}: {left!r} != {right!r}"
    if isinstance(left, list):
        if len(left) != len(right):
            return f"{prefix}: length {len(left)} != {len(right)}"
        for index, (left_item, right_item) in enumerate(zip(left, right)):
            nested = first_diff(left_item, right_item, f"{prefix}[{index}]")
            if nested:
                return nested
        return None
    keys = sorted(set(left) | set(right))
    for key in keys:
        if key not in left:
            return f"{prefix}.{key}: missing on actual"
        if key not in right:
            return f"{prefix}.{key}: missing on expected"
        nested = first_diff(left[key], right[key], f"{prefix}.{key}")
        if nested:
            return nested
    return None


def verify_snapshot(date: str) -> None:
    completed = subprocess.run(
        [sys.executable, str(ROOT / "models" / "mlb" / "cartridges" / "MLB-RP36" / "verify_snapshot.py"), "--date", date],
        cwd=ROOT,
        text=True,
        capture_output=True,
        check=False,
    )
    if completed.returncode:
        raise RuntimeError((completed.stdout + completed.stderr).strip())


def main() -> int:
    args = parse_args()
    date = args.date
    run_dir = RUN_ROOT / date
    run = read_json(run_dir / "run.json")
    snapshot = read_json(run_dir / "snapshot.json")
    actual_snapshot = build_snapshot(date)
    diff = first_diff(actual_snapshot, snapshot)
    if diff:
        raise RuntimeError(f"MLB-RP36 snapshot mismatch: {diff}")

    verify_snapshot(date)

    print(json.dumps({
        "runId": run["runId"],
        "status": "checked",
        "sourceFiles": run.get("sourceFiles"),
        "inputs": run.get("inputs"),
        "outputs": run.get("outputs"),
        "artifactHash": run["artifactHash"],
        "artifactSummary": run["artifactSummary"],
    }, indent=2))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(error, file=sys.stderr)
        raise SystemExit(1)

File: models/mlb/cartridges/MLB-M1/check_run.mjs
import fs from 'node:fs/promises'
import path from 'node:path'
import { buildM1Snapshot } from './snapshot.mjs'

const rootDir = path.resolve(import.meta.dirname, '..', '..', '..', '..')

const parseArgs = () => {
  const args = process.argv.slice(2)
  const options = { date: '' }
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === '--date') {
      options.date = args[index + 1]
      index += 1
    }
  }
  if (!/^\d{4}-\d{2}-\d{2}$/.test(options.date)) throw new Error('Pass --date YYYY-MM-DD')
  return options
}

const readJson = async (relativePath) => JSON.parse(await fs.readFile(path.resolve(rootDir, relativePath), 'utf8'))

const writeJson = async (relativePath, payload) => {
  const absolute = path.resolve(rootDir, relativePath)
  await fs.mkdir(path.dirname(absolute), { recursive: true })
  await fs.writeFile(absolute, `${JSON.stringify(payload, null, 2)}\n`)
}

const firstDiff = (left, right, prefix = '$') => {
  if (Object.is(left, right)) return null
  if (typeof left !== typeof right) return `${prefix}: type ${typeof left} !== ${typeof right}`
  if (left === null || right === null || typeof left !== 'object') return `${prefix}: ${JSON.stringify(left)} !== ${JSON.stringify(right)}`
  if (Array.isArray(left) !== Array.isArray(right)) return `${prefix}: array mismatch`
  if (Array.isArray(left)) {
    if (left.length !== right.length) return `${prefix}: length ${left.length} !== ${right.length}`
    for (let index = 0; index < left.length; index += 1) {
      const nested = firstDiff(left[index], right[index], `${prefix}[${index}]`)
      if (nested) return nested
    }
    return null
  }
  const keys = Array.from(new Set([...Object.keys(left), ...Object.keys(right)])).sort()
  for (const key of keys) {
    if (!(key in left)) return `${prefix}.${key}: missing on actual`
    if (!(key in right)) return `${prefix}.${key}: missing on expected`
    const nested = firstDiff(left[key], right[key], `${prefix}.${key}`)
    if (nested) return nested
  }
  return null
}

const main = async () => {
  const { date } = parseArgs()
  const runDir = `data-private/model-runs/mlb/MLB-M1/${date}`
  const run = await readJson(`${runDir}/run.json`)
  const snapshot = await readJson(`${runDir}/snapshot.json`)
  const actualSnapshot = await buildM1Snapshot({ date })
  const diff = firstDiff(actualSnapshot, snapshot)
  if (diff) {
    const actualPath = `${runDir}/snapshot.actual.json`
    await writeJson(actualPath, actualSnapshot)
    throw new Error(`MLB-M1 run snapshot mismatch: ${diff}. Actual written to ${actualPath}`)
  }
  console.log(JSON.stringify({
    runId: run.runId,
    status: 'checked',
    sourceFiles: run.sourceFiles,
    inputs: run.inputs,
    outputs: run.outputs,
    snapshotHash: run.snapshotHash,
    artifactSummary: run.artifactSummary
  }, null, 2))
}

main().catch((error) => {
  console.error(error.message || error)
  process.exitCode = 1
})
